src/lib/crypto/kimchi_backend/gadgets/ecdsa.ml

```diff
@@ -0,0 +1,1132 @@
+    if use_precomputed_gen_doubles then Some curve.doubles else None
+  in
+  let u1_point =
+    Ec_group.scalar_mul
+      (module Circuit)
+      base_checks curve ?doubles:curve_doubles u1_bits curve.gen
+  in
+
+  (* Bounds 5: Generator is gadget input (public parameter)
+   *           Initial accumulator is gadget input (checked externally or public parameter)
+   *           Result bound check for u1_point below.
+   *)
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x u1_point ;
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y u1_point ;
+
+  (* C6: Constrain scalar multiplication u2P *)
+  let u2_point =
+    Ec_group.scalar_mul
+      (module Circuit)
+      base_checks curve ?doubles u2_bits pubkey
+  in
+
+  (* Bounds 6: Pubkey is gadget input (checked externally)
+   *           Initial accumulator is gadget input (checked externally or public parameter)
+   *           Result bound check for u2_point below.
+   *)
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x u2_point ;
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y u2_point ;
+
+  (* C7: R = u1G + u2P *)
+  let result =
+    Ec_group.add (module Circuit) base_checks curve u1_point u2_point
+  in
+
+  (* Bounds 7: Left and right inputs checked by (Bounds 5) and (Bounds 6)
+   *           Result bound is bound checked below
+   *)
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x result ;
+  Foreign_field.External_checks.append_bound_check base_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y result ;
+
+  (* Constrain that r = Rx (mod n), where n is the scalar field modulus
+   *
+   *   Note: The scalar field modulus (curve.order) may be greater or smaller than
+   *         the base field modulus (curve.modulus)
+   *
+   *           curve.order > curve.modulus => Rx = 0 * n + Rx
+   *
+   *           curve.order < curve.modulus  => Rx = q * n + Rx'
+   *
+   *  Thus, to check for congruence we need to compute the modular reduction of Rx and
+   *  assert that it equals r.
+   *
+   *  Since we may want to target applications where the scalar field is much smaller
+   *  than the base field, so we cannot make any assumptions about the ratio between
+   *  these moduli, so we will constrain Rx = q * n + Rx' using the foreign field
+   *  multiplication gadget, rather than just constraining Rx + 0 with our foreign
+   *  field addition gadget.
+   *
+   *  As we are reducing Rx modulo n, we are performing foreign field arithmetic modulo n.
+   *  However, the multiplicand n above is not a valid foreign field element in [0, n - 1].
+   *  To be safe we must constrain Rx = q * (n - 1) + q + Rx  modulo n.
```

Comment:
??
the Rx on the right should be a Rx' no?

Reply:
Correct - typo.